Commit ID: 
github.com/COMPSCI-383-Fall2024/assignment-5-judging-flowers-<redacted> 

In [48]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression 
from sklearn.metrics import accuracy_score, confusion_matrix
import numpy as np
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely or locally)
dataset = pd.read_csv("iris.csv")
# Output the first 15 rows of the data
dataset.head(15)

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa
5,5.4,3.9,1.7,0.4,setosa
6,4.6,3.4,1.4,0.3,setosa
7,5.0,3.4,1.5,0.2,setosa
8,4.4,2.9,1.4,0.2,setosa
9,4.9,3.1,1.5,0.1,setosa


In [3]:
# Display a summary of the table information (number of datapoints, etc.)
dataset.describe(include='all')

,sepal_length,sepal_width,petal_length,petal_width,species
count,150.000000,150.000000,150.000000,150.000000,150
unique,NaN,NaN,NaN,NaN,3
top,NaN,NaN,NaN,NaN,setosa
freq,NaN,NaN,NaN,NaN,50
mean,5.843333,3.054000,3.758667,1.198667,NaN
std,0.828066,0.433594,1.764420,0.763161,NaN
min,4.300000,2.000000,1.000000,0.100000,NaN
25%,5.100000,2.800000,1.600000,0.300000,NaN
50%,5.800000,3.000000,4.350000,1.300000,NaN
75%,6.400000,3.300000,5.100000,1.800000,NaN


## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

We've got 4 features that describe different measurements of the face of the flower. The label describes what species the datapoint belongs to. The way I have output the dataset shows the classes are the species and these are labeled as "setosa", "versicolor", and "virginica". In the way this data is displayed, the labels are the actual classes. The features we are working with are: sepal_length, sepal_width, petal_length, petal_width. 

# Part 2: Split the dataset into train and test

In [4]:
# Take the dataset and split it into our features (X) and label (y)
x = dataset.drop('species', axis=1)
y = dataset['species']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=5)

# Part 3: Logistic Regression

In [5]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression()
model.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
probabilities = model.predict_proba([[5.1, 3.5, 1.4, 0.2]])
probabilities

/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


array([[9.80825315e-01, 1.91746427e-02, 4.24260768e-08]])

In [6]:
# iii. Report on the score for Logistic regression model, what does the score measure?
score = model.score(X_test, y_test)
score

0.9333333333333333

The score refers how well the model predicts the correct labeling on iterations of the testing data. A score of 1 indicates that the model predicted all instances of the data correct. If the score is 0, 0% of the predicts of label values given a feature were correct. This score only indicates the accuracy of the model on this dataset, the test dataset we have given it. 

In [16]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
model.coef_, model.intercept_

(array([[-0.44361636,  0.90588186, -2.3988224 , -1.03602713],
        [ 0.67046737, -0.47196691, -0.14771592, -1.08109961],
        [-0.22685101, -0.43391495,  2.54653832,  2.11712674]]),
 array([  9.62106283,   1.6843663 , -11.30542913]))

Decision Boundry for setosa
$0 = 9.62 -0.44x_1 + 0.91x_2 -2.40x_3 -1.04x_4 $
Decision Boundry for versicolor
$0 = 1.68 + 0.67x_1 -0.47x_2 -0.15x_3 -1.08x_4 $
Decision Boundry for virginica
$0 = -11.31 -0.23x_1 -0.43x_2 +2.55x_3 +2.12x_4 $

# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
SVCmodel = SVC(kernel = 'rbf', random_state=42, probability = True).fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
SVCmodel.predict_proba([[5.1, 3.5, 1.4, 0.2]])

/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


array([[0.9757053 , 0.01539077, 0.00890393]])

In [24]:
# iii. Report on the score for the SVM, what does the score measure?
score = SVCmodel.score(X_test, y_test)
score

0.9333333333333333

The score tells you how often the SVC model accurately predicts the label for the test dataset. This model correctly predicted the test dataset 93% of the time.  

# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
NNmodel = MLPClassifier(hidden_layer_sizes=(10, 10), max_iter=1000, random_state=42)
NNmodel.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
NNmodel.predict_proba([[5.1, 3.5, 1.4, 0.2]])

/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


array([[9.97067830e-01, 2.93175214e-03, 4.17437223e-07]])

In [30]:
# iii. Report on the score for the Neural Network, what does the score measure?
score = NNmodel.score(X_test, y_test)
score

0.8

The score tells you how often the neural network model accurately predicts the label for the test dataset. This model correctly predicted the test dataset 80% of the time.  

In [ ]:
# iv: Experiment with different options for the neural network, report on your best configuration
NNmodel = MLPClassifier(hidden_layer_sizes=(30, 30), max_iter=1000, random_state=42)
NNmodel.fit(X_train, y_train)
NNmodel.score(X_test, y_test)

0.9333333333333333

The best configuration I could muster is shown above. The score indicates a 93.33% accuracy. 

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
Kmodel = KNeighborsClassifier(n_neighbors=3).fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
Kmodel.predict_proba([[5.1, 3.5, 1.4, 0.2]])

/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


array([[1., 0., 0.]])

In [51]:
# iii. Report on the score for kNN, what does the score measure?
model.score(X_test, y_test)

0.9333333333333333

As I've mentioned. The score refers to the accuracy of the model at predicting the correct labeling of the test dataset. This model accurately predicted the labeling of the test data 93% of the time. 

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The k-nearest neighbors had the best classification prediction for the data point (which was extracted from the inital dataset to be a setosa flower). This model predicted the probability it was setosa with 1, and the other classes 0. I thought this was prettu interesting how it was able to say definatively this was the class. The score of this model was still only 93% which is tied for the most accurate model with the neural, SVM, and logistic regression. I thought it was interesting that all these models had the same score. I'm thinking maybe this occured because the data is relatively small, they all used the same test and train, maybe some characteristic of that data partition allowed for these different models to be trained with similar accuracy. I am not saying that this will always be the case, I know that the same data != the same accuracy. I am just speculating here because to be honest I really do not know. 